API/common/context.py

```diff
@@ -0,0 +1,221 @@
+            self.embeddingUrl = os.environ.get("EMBEDDING_MODEL_URL")
+        else:
+            self.embeddingUrl = embeddingUrl
+        
+    def toDict(self):
+        return {
+            "apiUrl": self.apiUrl,
+            "apiVersion": self.apiVersion,
+            "gptEngine": self.gptEngine,
+            "embeddingUrl": self.embeddingUrl
+        }
+
+
+class Context():
+    statistics = None
+    originalCall: DependencyCall = DependencyCall("OriginalCall")
+    functionOperationId: str
+    functionInvocationId: str
+    applicationInsightsId: str
+    openaiConfig: OpenAIConfig = None
+    tracer = None
+    method = None
+    endpoint = None
+
+    def __init__(self, functionOperationId: str, functionInvocationId: str, openaiConfig: OpenAIConfig) -> None:
+        self.functionOperationId = functionOperationId
+        self.functionInvocationId = functionInvocationId
+        self.statistics = ContextStatistics()
+        self.openaiConfig = openaiConfig
+
+    def calculateUsage(self):
+        totalUsage = {}
+        for call in self.statistics.callGraph:
+            usage = call.usage
+            if usage is None:
+                continue
+            model = usage.get('model', None)
+            if model is None:
+                model = "None"
+            if model not in totalUsage:
+                totalUsage[model] = {
+                    'completion_tokens': 0,
+                    'prompt_tokens': 0,
+                    'total_tokens': 0,
+                }
+            modelUsage = totalUsage[model]
+            modelUsage['completion_tokens'] += usage.get('completion_tokens', 0)
+            modelUsage['prompt_tokens'] += usage.get('prompt_tokens', 0)
+            modelUsage['total_tokens'] += usage.get('total_tokens', 0)
+        self.originalCall.usage = totalUsage
+
+    def toDict(self):
+        return {
+            "method": self.method,
+            "endpoint": self.endpoint,
+            "statistics": self.statistics.toDict(),
+            "originalCall": self.originalCall.toDict(),
+            "functionOperationId": self.functionOperationId,
+            "functionInvocationId": self.functionInvocationId,
+            "applicationInsightsId": self.applicationInsightsId
+        }
+
+def log_dependency_call(callName):
+    def decorate(func):
+        @wraps(func)
+        def innerFunc(*args, **kwargs):
+            context = args[0].custom_context
+            dependencyCall = DependencyCall(callName)
+            dependencyCall.start()
+            try:
+                tracer = context.tracer
+                with tracer.span(name=callName) as span:
+                    span.add_attribute("name", callName)
+                    response = func(*args, **kwargs)
+                    if 'usage' in response:
+                        dependencyCall.usage = response['usage']
+                        for key, value in response['usage'].items():
+                            span.add_attribute(f"{key}", value)
```

Comment:
After logging `response['usage']`, could we del `usage` from `responese` to avoid exposing unnecessary token consumption information?

Reply:
I removed other fields [here](https://github.com/hackathon-cli-recommendation/cli-recommendation/pull/52/files#diff-dbaa0dabb719cc69c522dbeaf0261ed7bc505392d64181cc8635f3c967ea5ab5R120) and only kept content. What do you think?
![image](https://github.com/hackathon-cli-recommendation/cli-recommendation/assets/18628534/255c85db-f199-4ffd-bb94-5def923096c0)